packages/compile-solidity/src/compilerSupplier/index.ts

```diff
@@ -0,0 +1,117 @@
+
+import { Docker, Local, Native, VersionRange } from "./loadingStrategies";
+
+const defaultSolcVersion = "0.5.16";
+
+export class CompilerSupplier {
+  private events: any;
+  private version: string;
+  private docker: boolean;
+  private compilerRoots: string[];
+  private strategyOptions: Partial<{
+    version: string;
+    docker: boolean;
+    compilerRoots: string[];
+    events: any; // represents a @truffle/events instance, which lacks types
+    spawn: {
+      maxBuffer: number;
+    };
+  }>;
+
+  constructor({ events, solcConfig }) {
+    const { version, docker, compilerRoots, spawn } = solcConfig;
+    this.events = events;
+    this.version = version ? version : defaultSolcVersion;
+    this.docker = docker;
+    this.compilerRoots = compilerRoots;
+    this.strategyOptions = {};
+    if (version) this.strategyOptions.version = this.version;
+    if (docker) this.strategyOptions.docker = compilerRoots;
+    if (compilerRoots) this.strategyOptions.compilerRoots = compilerRoots;
+    if (events) this.strategyOptions.events = events;
+    if (spawn) this.strategyOptions.spawn = spawn;
+  }
+
+  async load() {
+    const userSpecification = this.version;
+
+    let strategy;
+    const useDocker = this.docker;
+    const useNative = userSpecification === "native";
+    const useSpecifiedLocal =
+      userSpecification && this.fileExists(userSpecification);
+    const isValidVersionRange = semver.validRange(userSpecification);
+
+    if (useDocker) {
+      strategy = new Docker(this.strategyOptions);
+    } else if (useNative) {
+      strategy = new Native();
+    } else if (useSpecifiedLocal) {
+      strategy = new Local();
+    } else if (isValidVersionRange) {
+      strategy = new VersionRange(this.strategyOptions);
+    }
+
+    if (strategy) {
+      const solc = await strategy.load(userSpecification);
+      return { solc };
+    } else {
+      throw this.badInputError(userSpecification);
+    }
+  }
+
+  async list() {
+    const userSpecification = this.version;
+
+    let strategy;
+    const useDocker = this.docker;
+    const useNative = userSpecification === "native";
+    const useSpecifiedLocal =
+      userSpecification && this.fileExists(userSpecification);
+    const isValidVersionRange = semver.validRange(userSpecification);
+
+    if (useDocker) {
+      strategy = new Docker(this.strategyOptions);
+    } else if (useNative) {
+      strategy = new Native();
+    } else if (useSpecifiedLocal) {
+      strategy = new Local();
+    } else if (isValidVersionRange) {
+      strategy = new VersionRange(this.strategyOptions);
+    }
+
+    if (!strategy) {
+      throw this.badInputError(userSpecification);
+    }
+
+    if (!strategy.list) {
+      throw new Error(
+        `Cannot list versions for strategy ${strategy.constructor.name}`
+      );
+    }
+
+    return await strategy.list();
+  }
+
+  static getDefaultVersion() {
+    return defaultSolcVersion;
+  }
+
+  badInputError(userSpecification) {
```

Comment:
@haltman-at added a note [here](https://github.com/trufflesuite/truffle/pull/4222#discussion_r687276676) about this

Reply:
Ah, I see you've changed this. :)